### ISSUE TITLE
Spatial transformer implementation looks different with the original version in the paper.

### ISSUE THREAD

**Issue by iwyoo** [tensorflow/models#295]
Hi, guys. I found that this spatial transformer implementation of tensorflow repository has one difference with the original paper's. 

In https://github.com/tensorflow/models/blob/master/transformer/spatial_transformer.py
```python
            x0 = tf.cast(tf.floor(x), 'int32')
            x1 = x0 + 1
            y0 = tf.cast(tf.floor(y), 'int32')
            y1 = y0 + 1
            ...
            Ia = tf.gather(im_flat, idx_a)
            Ib = tf.gather(im_flat, idx_b)
            Ic = tf.gather(im_flat, idx_c)
            Id = tf.gather(im_flat, idx_d)
```

This implementation just applies type-casting/gathering operator for interpolation (by index access), but it doesn't seem that it is differentiable as in the paper. Shouldn't it be bilinear interpolation with maximum operation?

Please check about it, and notice me what you think.
Inwan

**Comment by michaelisard** [comment]
@martinwicke do you know who owns this model?

**Comment by martinwicke** [comment]
@daviddao contributed it. David, can you comment?

**Comment by ppwwyyxx** [comment]
Bilinear sampling happens here:
```python
            wa = tf.expand_dims(((x1_f-x) * (y1_f-y)), 1)
            wb = tf.expand_dims(((x1_f-x) * (y-y0_f)), 1)
            wc = tf.expand_dims(((x-x0_f) * (y1_f-y)), 1)
            wd = tf.expand_dims(((x-x0_f) * (y-y0_f)), 1)
```
I also had an [implementation](https://github.com/ppwwyyxx/tensorpack/blob/master/tensorpack/models/image_sample.py#L42) of bilinear sampling simliar to this.

The formulation in the paper is just another equivalent way of writing this.

**Comment by iwyoo** [comment]
Thank you for your comment! However, I still can't find the maximum operator that makes this method (sub)differentiable. Is there any part I'm missing?

**Comment by daviddao** [comment]
Hi @iwyoo, 

I am not sure, if I correctly get your point.
@ppwwyyxx correctly pointed to the bilinear sampling part. 
Also see [this implementation](http://stackoverflow.com/questions/34902782/interpolated-sampling-of-points-in-an-image-with-tensorflow) as example.

In the original STN paper, the maximum operator inside the bilinear sampling kernel is used as zero clipping (referring to (5) in Jaderberg _et al._). This is done here:

```python
            x0 = tf.clip_by_value(x0, zero, max_x)
            x1 = tf.clip_by_value(x1, zero, max_x)
            y0 = tf.clip_by_value(y0, zero, max_y)
            y1 = tf.clip_by_value(y1, zero, max_y)
```

**Comment by ppwwyyxx** [comment]
@iwyoo One mathematical formula can be written in many different ways.
@daviddao The maximum is not used to clip the coordinate in the range. It is simply used to find out the coordinates to interpolate with.

For example, `\sum_m^W max(0, 1-\abs{x - m})`  is non-zero if and only if:
 `m=floor(x)` or `m=ceiling(x)` (assuming x is not an integer).
This is equivalent to the floor/ceiling part in our code, for example [here](https://github.com/tensorflow/models/blob/master/transformer/spatial_transformer.py#L83).

It is not the maximum operator that makes bilinear sample differentiable. Bilinear sample __itself__ is differentiable.

**Comment by daviddao** [comment]
@ppwwyyxx You are right! Thanks I got confused! @iwyoo Please go with the answer above.

**Comment by iwyoo** [comment]
Thank you for your kind comments, @daviddao and @ppwwyyxx .
 One thing I last wonder is the index access by tf.gather. In the original paper, the method doesn't access image pixel value by index access but by sum of all interpolated pixels. ((3)-(5) in Jaderberg et al.)
 (6), (7) of the reference can be derived because it is based on the exhausted sum operation over the pixels. Is it still differentiable even if we use tf.gather instead of sum operation?

**Comment by ppwwyyxx** [comment]
As I mentioned, although they summed up the pixels over the image, the only non-zero part would be the nearby coordinates to interpolate with. Mathematically they are the same.
`tf.gather` has a zero-gradient w.r.t to index, and a one-gradient w.r.t whatever is gathered. So the gradient of the whole operator is nicely taken care of by automatic differentiation.

**Comment by iwyoo** [comment]
If tf.gather has a zero-gradient w.r.t. index, the generated grid coordinate also has zero-gradient. It means that the localisation of this method can't be trained. This implementation do transform input image by given affine transform(theta), but it doesn't seem that it can't learn localisation (as in the original paper.)

**Comment by ppwwyyxx** [comment]
the weight in interpolation has nonzero gradient wrt  coordinates. that's where the gradients propogate.

**Comment by iwyoo** [comment]
Oh, now I fully understand it. I'll close this issue. Thank you all guys!

**Comment by iperov** [comment]
I confirm
tf.gather in flat indexes has wrong gradient than described in paper, 
where gradients have value -1 0 1 depends on 2D direction for every x,y coord axis
